File: data-pipeline-libraries-java/data-pipeline-gcp-observability-java/src/main/java/com/enrichmeai/culvert/gcp/observability/CloudTraceObservabilityHook.java
package com.enrichmeai.culvert.gcp.observability;

import com.enrichmeai.culvert.contracts.ObservabilityHook;
import io.opentelemetry.api.OpenTelemetry;
import io.opentelemetry.api.common.AttributeKey;
import io.opentelemetry.api.metrics.DoubleHistogram;
import io.opentelemetry.api.metrics.LongCounter;
import io.opentelemetry.api.metrics.Meter;
import io.opentelemetry.api.trace.Tracer;
import io.opentelemetry.context.Scope;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Map;
import java.util.Objects;
import java.util.concurrent.ConcurrentHashMap;

/**
 * {@link ObservabilityHook} implementation backed by OpenTelemetry whose
 * tracer / meter providers export to Google Cloud Trace and Google Cloud
 * Monitoring.
 *
 * <p>Wiring is decoupled from this class: callers construct an
 * {@link OpenTelemetry} instance whose {@code SdkTracerProvider} has a
 * {@code TraceExporter} from the
 * {@code com.google.cloud.opentelemetry:exporter-trace} artifact registered
 * (typically via a {@code BatchSpanProcessor}). This class wraps the
 * resulting {@link Tracer} and {@link Meter} and bridges them to the Culvert
 * {@link ObservabilityHook} surface.
 *
 * <h2>Construction</h2>
 * <ul>
 *   <li>{@link #CloudTraceObservabilityHook()} — no-arg, for {@link java.util.ServiceLoader}
 *       discovery (T12.6). Wraps {@link io.opentelemetry.api.GlobalOpenTelemetry#get()} with a
 *       fixed instrumentation scope ({@value #DEFAULT_INSTRUMENTATION_SCOPE}). Consistent with
 *       the class design: wiring the OTel exporter to Cloud Trace is the caller's responsibility;
 *       if the global OTel SDK is configured with a GCP trace exporter, this hook transparently
 *       emits to Cloud Trace. If only the no-op global is installed, spans are discarded — this
 *       is the expected fallback when the module is on the classpath but no OTel SDK is wired.
 *       This constructor never throws — {@link io.opentelemetry.api.GlobalOpenTelemetry#get()}
 *       always returns a non-null instance.</li>
 *   <li>{@link #CloudTraceObservabilityHook(OpenTelemetry, String)} — primary;
 *       caller supplies an OTel instance + an instrumentation scope name
 *       (typically the pipeline name).</li>
 *   <li>{@link #CloudTraceObservabilityHook(Tracer, Meter)} — for tests; wraps
 *       an already-resolved {@link Tracer} and {@link Meter} so callers can
 *       inject mocks directly.</li>
 * </ul>
 *
 * <h2>Bridging the contract</h2>
 * <ul>
 *   <li>{@code counter}/{@code gauge}/{@code histogram} → OTel {@link Meter}
 *       instruments, lazily created and cached by name.</li>
 *   <li>{@code log} → SLF4J at the requested level. The {@code fields} map is
 *       rendered as {@code key=value} pairs appended to the message (SLF4J
 *       has no structured-field API in core, so this keeps the log line
 *       self-contained without forcing a logback / MDC dependency).</li>
 *   <li>{@code span} → OTel {@link Tracer}; the returned {@link Span} wraps
 *       the OTel {@code Span}+{@code Scope} pair and closes both on
 *       {@link Span#close()}.</li>
 * </ul>
 *
 * <p>This class does <strong>not</strong> implement {@link AutoCloseable}:
 * the wrapped {@link OpenTelemetry} interface is not closeable; lifecycle
 * (flushing the BatchSpanProcessor on shutdown) belongs to whoever built
 * the SDK. Mirrors the {@code BigQueryWarehouse} "AutoCloseable only when
 * the wrapped client supports it" rule.
 *
 * <p>Sprint-2 deliverable for issue #24; no-arg ctor added Sprint-12 T12.6 (issue #91).
 */
public final class CloudTraceObservabilityHook implements ObservabilityHook {

    private static final Logger LOG = LoggerFactory.getLogger(CloudTraceObservabilityHook.class);

    /**
     * Default instrumentation scope used by the no-arg constructor.
     *
     * <p>An instrumentation scope identifies the library or component that produces spans.
     * When wired via ServiceLoader, the scope is fixed to this constant; callers that need
     * a pipeline-specific scope should use the {@link #CloudTraceObservabilityHook(OpenTelemetry, String)}
     * constructor directly.
     */
    static final String DEFAULT_INSTRUMENTATION_SCOPE = "com.enrichmeai.culvert";

    private final Tracer tracer;
    private final Meter meter;

    // Instrument caches — OTel meters require us to register instruments
    // by name; reusing instances avoids re-registration cost and keeps
    // metric IDs stable across calls.
    private final Map<String, LongCounter> counters = new ConcurrentHashMap<>();
    private final Map<String, DoubleHistogram> histograms = new ConcurrentHashMap<>();
    private final Map<String, DoubleHistogram> gauges = new ConcurrentHashMap<>();

    /**
     * No-arg constructor for {@link java.util.ServiceLoader} discovery (T12.6).
     *
     * <p>Wraps {@link io.opentelemetry.api.GlobalOpenTelemetry#get()} with the fixed
     * instrumentation scope {@value #DEFAULT_INSTRUMENTATION_SCOPE}. Consistent with this
     * class's stated design: wiring the OTel SDK to export to Cloud Trace is the caller's
     * responsibility (via a {@code TraceExporter} + {@code BatchSpanProcessor} on the
     * {@code SdkTracerProvider}). If the global OTel is configured with such an exporter,
     * spans are emitted to Cloud Trace. If the no-op global is installed (e.g. when no
     * OTel SDK is on the classpath), spans are discarded — this is the expected graceful
     * fallback.
     *
     * <p>This constructor <strong>never throws</strong>:
     * {@link io.opentelemetry.api.GlobalOpenTelemetry#get()} always returns a non-null
     * instance, and {@link OpenTelemetry#getTracer}/{@link OpenTelemetry#getMeter} are
     * similarly non-throwing. This makes the class unconditionally instantiable by
     * ServiceLoader — the SPI registration is now real (T12.6).
     */
    public CloudTraceObservabilityHook() {
        this(io.opentelemetry.api.GlobalOpenTelemetry.get(), DEFAULT_INSTRUMENTATION_SCOPE);
    }

    /**
     * Primary constructor.
     *
     * @param otel               Configured OpenTelemetry instance whose tracer
     *                           provider exports to Cloud Trace. Required.
     * @param instrumentationScope Name passed to {@code getTracer} /
     *                           {@code getMeter} — typically the pipeline
     *                           name. Required.
     * @throws NullPointerException if either argument is null.
     */
    public CloudTraceObservabilityHook(OpenTelemetry otel, String instrumentationScope) {
        Objects.requireNonNull(otel, "otel must not be null");
        Objects.requireNonNull(instrumentationScope, "instrumentationScope must not be null");
        this.tracer = otel.getTracer(instrumentationScope);
        this.meter = otel.getMeter(instrumentationScope);
    }

    /**
     * Test-friendly constructor that takes pre-resolved {@link Tracer} and
     * {@link Meter}. Both are required.
     *
     * @throws NullPointerException if either argument is null.
     */
    public CloudTraceObservabilityHook(Tracer tracer, Meter meter) {
        this.tracer = Objects.requireNonNull(tracer, "tracer must not be null");
        this.meter = Objects.requireNonNull(meter, "meter must not be null");
    }

    @Override
    public void counter(String name, long value, Map<String, String> tags) {
        Objects.requireNonNull(name, "name");
        LongCounter c = counters.computeIfAbsent(name, n -> meter.counterBuilder(n).build());
        c.add(value, toAttributes(tags));
    }

    @Override
    public void gauge(String name, double value, Map<String, String> tags) {
        Objects.requireNonNull(name, "name");
        // OpenTelemetry's "synchronous gauge" landed in 1.40+. For 1.38 we
        // surface gauges as a single-shot histogram observation — this keeps
        // the value visible in Cloud Monitoring without requiring async
        // callbacks. When the SDK is bumped to 1.40+ this should switch to
        // meter.gaugeBuilder(name).build().
        DoubleHistogram g = gauges.computeIfAbsent(
                name, n -> meter.histogramBuilder(n).build());
        g.record(value, toAttributes(tags));
    }

    @Override
    public void histogram(String name, double value, Map<String, String> tags) {
        Objects.requireNonNull(name, "name");
        DoubleHistogram h = histograms.computeIfAbsent(
                name, n -> meter.histogramBuilder(n).build());
        h.record(value, toAttributes(tags));
    }

    @Override
    public void log(String level, String message, Map<String, Object> fields) {
        Objects.requireNonNull(level, "level");
        Objects.requireNonNull(message, "message");
        String rendered = renderFields(message, fields);
        String upper = level.toUpperCase(java.util.Locale.ROOT);
        switch (upper) {
            case "DEBUG" -> LOG.debug(rendered);
            case "INFO" -> LOG.info(rendered);
            case "WARN" -> LOG.warn(rendered);
            case "ERROR" -> LOG.error(rendered);
            default -> LOG.info(rendered);
        }
    }

    @Override
    public Span span(String name) {
        Objects.requireNonNull(name, "name");
        io.opentelemetry.api.trace.Span otelSpan = tracer.spanBuilder(name).startSpan();
        Scope scope = otelSpan.makeCurrent();
        return new OtelSpanAdapter(otelSpan, scope);
    }

    // --- helpers ----------------------------------------------------------

    private static io.opentelemetry.api.common.Attributes toAttributes(Map<String, String> tags) {
        if (tags == null || tags.isEmpty()) {
            return io.opentelemetry.api.common.Attributes.empty();
        }
        var builder = io.opentelemetry.api.common.Attributes.builder();
        for (Map.Entry<String, String> e : tags.entrySet()) {
            if (e.getKey() != null && e.getValue() != null) {
                builder.put(AttributeKey.stringKey(e.getKey()), e.getValue());
            }
        }
        return builder.build();
    }

    private static String renderFields(String message, Map<String, Object> fields) {
        if (fields == null || fields.isEmpty()) {
            return message;
        }
        StringBuilder sb = new StringBuilder(message);
        sb.append(' ');
        boolean first = true;
        for (Map.Entry<String, Object> e : fields.entrySet()) {
            if (!first) {
                sb.append(' ');
            }
            sb.append(e.getKey()).append('=').append(e.getValue());
            first = false;
        }
        return sb.toString();
    }

    /**
     * Adapter from Culvert {@link Span} to OpenTelemetry's
     * {@code Span}+{@code Scope} pair. Idempotent on {@link #close()} — a
     * double-close is a no-op rather than throwing, matching the
     * try-with-resources expectations of the contract.
     */
    static final class OtelSpanAdapter implements Span {
        private final io.opentelemetry.api.trace.Span span;
        private final Scope scope;
        private boolean closed;

        OtelSpanAdapter(io.opentelemetry.api.trace.Span span, Scope scope) {
            this.span = span;
            this.scope = scope;
        }

        @Override
        public void setAttribute(String key, String value) {
            Objects.requireNonNull(key, "key");
            Objects.requireNonNull(value, "value");
            span.setAttribute(key, value);
        }

        @Override
        public void recordException(Throwable t) {
            Objects.requireNonNull(t, "t");
            span.recordException(t);
        }

        @Override
        public void close() {
            if (closed) {
                return;
            }
            closed = true;
            try {
                scope.close();
            } finally {
                span.end();
            }
        }
    }
}

File: data-pipeline-libraries/data-pipeline-gcp-observability/src/data_pipeline_gcp_observability/cloud_trace_observability_hook.py
"""CloudTraceObservabilityHook — ObservabilityHook backed by OpenTelemetry.

Java sibling:
  data-pipeline-libraries-java/data-pipeline-gcp-observability-java/src/main/java/
  com/enrichmeai/culvert/gcp/observability/CloudTraceObservabilityHook.java

The OTel SDK and exporter are injected by the caller; this class only wraps
them as the Culvert ObservabilityHook Protocol surface.  All SDK imports are
performed lazily so the module can be imported even when the OTel SDK is not
installed (e.g. in unit test environments that only install mocks).

Sprint-19 / T19.2 — issue #125.
"""

from __future__ import annotations

import logging
from contextlib import contextmanager
from typing import Any, Generator, Mapping

logger = logging.getLogger(__name__)

# Sentinel for empty tag mappings — avoids dict construction on every call.
_NO_TAGS: Mapping[str, str] = {}


class CloudTraceObservabilityHook:
    """ObservabilityHook implementation backed by OpenTelemetry.

    Mirrors Java ``CloudTraceObservabilityHook`` (file line 71).

    Callers construct an OpenTelemetry instance whose tracer / meter providers
    export to Google Cloud Trace / Cloud Monitoring and pass it here.  This
    class only wraps the resulting tracer and meter and bridges them to the
    Culvert ``ObservabilityHook`` Protocol.

    Construction
    ------------
    - ``CloudTraceObservabilityHook(tracer, meter)`` — primary; inject
      already-resolved tracer and meter (OTel SDK objects or mocks).
    - ``CloudTraceObservabilityHook.from_otel(otel, scope)`` — convenience
      class-method; mirrors Java ``(OpenTelemetry, String)`` constructor
      (line 127).  Requires the OTel SDK to be installed.
    - ``CloudTraceObservabilityHook()`` — no-arg; wraps the OTel global SDK
      (mirrors Java no-arg ctor, line 113).  Requires the OTel SDK.

    The no-arg and ``from_otel`` paths perform lazy SDK imports so that the
    class can be imported even when the SDK is absent.

    Instrument caches (counters, histograms) are ``dict``-backed; OTel
    instruments are keyed by name so re-registration is avoided, mirroring
    Java ``ConcurrentHashMap`` caches (line 91–93).
    """

    DEFAULT_INSTRUMENTATION_SCOPE = "com.enrichmeai.culvert"

    def __init__(self, tracer: Any, meter: Any) -> None:
        """Inject an already-resolved OTel tracer and meter.

        Both are required.  Pass ``MagicMock()`` instances in tests.

        Args:
            tracer: OTel Tracer (or mock).
            meter:  OTel Meter (or mock).

        Raises:
            TypeError: if either argument is None.
        """
        if tracer is None:
            raise TypeError("tracer must not be None")
        if meter is None:
            raise TypeError("meter must not be None")
        self._tracer = tracer
        self._meter = meter
        self._counters: dict[str, Any] = {}
        self._histograms: dict[str, Any] = {}
        self._gauges: dict[str, Any] = {}

    @classmethod
    def from_otel(cls, otel: Any, scope: str) -> "CloudTraceObservabilityHook":
        """Build from an ``OpenTelemetry`` instance and an instrumentation scope.

        Mirrors Java primary constructor (line 127).

        Args:
            otel:  Configured OpenTelemetry instance.
            scope: Instrumentation scope name (typically the pipeline name).

        Raises:
            TypeError:     if either argument is None.
            ImportError:   if ``opentelemetry-api`` is not installed.
        """
        if otel is None:
            raise TypeError("otel must not be None")
        if scope is None:
            raise TypeError("scope must not be None")
        return cls(otel.get_tracer(scope), otel.get_meter(scope))

    @classmethod
    def default(cls) -> "CloudTraceObservabilityHook":
        """Build from the OTel global SDK.

        Mirrors Java no-arg constructor (line 113).

        Raises:
            ImportError: if ``opentelemetry-api`` is not installed.
        """
        try:
            from opentelemetry import trace, metrics  # type: ignore[import]
        except ImportError as exc:  # pragma: no cover
            raise ImportError(
                "opentelemetry-api is required for CloudTraceObservabilityHook.default(). "
                "Install it with: pip install opentelemetry-api"
            ) from exc
        tracer = trace.get_tracer(cls.DEFAULT_INSTRUMENTATION_SCOPE)
        meter = metrics.get_meter(cls.DEFAULT_INSTRUMENTATION_SCOPE)
        return cls(tracer, meter)

    # ------------------------------------------------------------------
    # ObservabilityHook Protocol
    # ------------------------------------------------------------------

    def counter(
        self,
        name: str,
        value: int = 1,
        tags: Mapping[str, str] = _NO_TAGS,
    ) -> None:
        """Increment a monotonic counter — mirrors Java counter() (line 146)."""
        if name is None:
            raise TypeError("name must not be None")
        instrument = self._counters.get(name)
        if instrument is None:
            instrument = self._meter.create_counter(name)
            self._counters[name] = instrument
        instrument.add(value, self._to_attributes(tags))

    def gauge(
        self,
        name: str,
        value: float,
        tags: Mapping[str, str] = _NO_TAGS,
    ) -> None:
        """Record a gauge value — mirrors Java gauge() (line 153)."""
        if name is None:
            raise TypeError("name must not be None")
        instrument = self._gauges.get(name)
        if instrument is None:
            # Use histogram as a gauge proxy (mirrors Java comment, line 158).
            instrument = self._meter.create_histogram(name)
            self._gauges[name] = instrument
        instrument.record(value, self._to_attributes(tags))

    def histogram(
        self,
        name: str,
        value: float,
        tags: Mapping[str, str] = _NO_TAGS,
    ) -> None:
        """Record a histogram observation — mirrors Java histogram() (line 166)."""
        if name is None:
            raise TypeError("name must not be None")
        instrument = self._histograms.get(name)
        if instrument is None:
            instrument = self._meter.create_histogram(name)
            self._histograms[name] = instrument
        instrument.record(value, self._to_attributes(tags))

    def log(self, level: str, message: str, **fields: Any) -> None:
        """Structured log — mirrors Java log() (line 174).

        ``level`` is mapped to the stdlib logging level (DEBUG/INFO/WARNING/ERROR).
        ``**fields`` are appended as ``key=value`` pairs on the log message,
        mirroring the Java ``renderFields`` helper (line 211).
        """
        if level is None:
            raise TypeError("level must not be None")
        if message is None:
            raise TypeError("message must not be None")
        rendered = self._render_fields(message, fields)
        upper = level.upper()
        if upper == "DEBUG":
            logger.debug(rendered)
        elif upper in ("WARNING", "WARN"):
            logger.warning(rendered)
        elif upper == "ERROR":
            logger.error(rendered)
        elif upper == "CRITICAL":
            logger.critical(rendered)
        else:
            logger.info(rendered)

    @contextmanager
    def span(self, name: str) -> Generator[Any, None, None]:
        """Open a tracing span — mirrors Java span() + OtelSpanAdapter (line 189, 234).

        Uses the injected tracer.  If the tracer is a mock the context manager
        still works because we guard every OTel call.

        Yields the underlying span object (or mock).
        """
        if name is None:
            raise TypeError("name must not be None")
        otel_span = self._tracer.start_span(name)
        try:
            yield otel_span
        except Exception as exc:
            try:
                otel_span.record_exception(exc)
            except Exception:  # noqa: BLE001
                pass
            raise
        finally:
            try:
                otel_span.end()
            except Exception:  # noqa: BLE001
                pass

    # ------------------------------------------------------------------
    # Helpers
    # ------------------------------------------------------------------

    @staticmethod
    def _to_attributes(tags: Mapping[str, str]) -> dict[str, str]:
        """Convert a tags mapping to a plain dict (OTel attribute map)."""
        if not tags:
            return {}
        return {k: v for k, v in tags.items() if k is not None and v is not None}

    @staticmethod
    def _render_fields(message: str, fields: dict[str, Any]) -> str:
        """Append ``key=value`` pairs to ``message`` — mirrors Java renderFields (line 211)."""
        if not fields:
            return message
        parts = " ".join(f"{k}={v}" for k, v in fields.items())
        return f"{message} {parts}"
